### PR TITLE
[ONNX] Support non-persistent buffers in FakeMode export

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107411
* #107247

Use `_extract_all_buffers_and_params` instead of `state_dict()` to get all paraters and buffers from models. This PR demonstrates a way to support non-persistent buffers in ONNX fake mode exporting.

NOTE: this PR doesn't address https://github.com/pytorch/pytorch/issues/107475

### CODE REVIEW

File: test/onnx/test_fx_to_onnx.py [hunk 1]
```diff
@@ -407,28 +407,7 @@ def forward(self, x):
                     fake_model, real_x, export_options=export_options
                 )
 
-    def test_fake_tensor_mode_huggingface_gpt2(self):
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx.py:410]: Have it in runtime

File: test/onnx/test_fx_to_onnx.py [hunk 2]
```diff
@@ -470,52 +450,6 @@ def test_fake_tensor_mode_huggingface_open_llama(self):
             onnx.checker.check_model(export_output.model_proto)
             onnx.shape_inference.infer_shapes(export_output.model_proto)
 
-    def test_fake_tensor_mode_huggingface_google_t5(self):
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx.py:473]: Have it in runtime

File: test/onnx/test_fx_to_onnx.py [hunk 3]
```diff
@@ -470,52 +450,6 @@ def test_fake_tensor_mode_huggingface_open_llama(self):
...
-                model,
-                input_ids=input_ids,
-                attention_mask=attention_mask,
-                decoder_input_ids=decoder_input_ids,
-                export_options=export_options,
-            )
-            onnx.checker.check_model(export_output.model_proto)
-            onnx.shape_inference.infer_shapes(export_output.model_proto)
-
-    def test_fake_tensor_mode_huggingface_openai_whisper(self):
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx.py:493]: Have it in runtime

File: test/onnx/test_fx_to_onnx.py (hunk 2 above)
Comment (wschin) [reply test/onnx/test_fx_to_onnx.py:473]: I think you meant `moved this test to `test_fx_to_onnx_with_onnxruntime.py`... Please keep comments clear.

**Comment by wschin** [comment]
What are the reasons not to fix https://github.com/pytorch/pytorch/issues/107475 in this PR?

**Comment by titaiwangms** [comment]
> What are the reasons not to fix #107475 in this PR?

Simply just I don't know how to get constant tensors in nn.Modules. Do you have any idea to do it?

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 4]
```diff
@@ -891,15 +906,13 @@ def _test_fake_tensor_mode_exporter(
...
             # The file will be loaded via .load_state_dict(...)
-            state_dict = real_model.state_dict()
+            state_dict = _extract_all_buffers_and_params(real_model)
             torch.save(state_dict, tmp_checkpoint_file.name)
 
             with torch.onnx.enable_fake_mode() as fake_context:
                 fake_args = create_args()
                 fake_kwargs = create_kwargs()
                 fake_model = create_model()
-                if load_checkpoint_during_init:
```
Comment (BowenBao) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:901]: I think the purpose of this flag was to cover both use case scenarios:
- Load model by `from_pretrained`.
- Load model by `load_state_dict` from checkpoint file.

Make sure both are still covered after your change.

**Review by BowenBao — COMMENTED** [review]
This is rather changing how test is written to get it pass... The root problem is still there that concrete non-persistent buffers and constants cannot be recovered due to they are not saved in checkpoint file by definition. This PR implies users need to preprocess their model checkpoint in order for fake mode export to work.

I think the right fix is to patch fake tensor mode under this context to keep track of mapping between fake tensors and their source concrete tensors. We need to be careful to only track those for non-persistent buffer and constants and leave out parameters and persistent buffers such that we don't defeat the purpose of fake mode export.

**Comment by titaiwangms** [comment]
> This is rather changing how test is written to get it pass... The root problem is still there that concrete non-persistent buffers and constants cannot be recovered due to they are not saved in checkpoint file by definition. This PR implies users need to preprocess their model checkpoint in order for fake mode export to work.
> 
> I think the right fix is to patch fake tensor mode under this context to keep track of mapping between fake tensors and their source concrete tensors. We need to be careful to only track those for non-persistent buffer and constants and leave out parameters and persistent buffers such that we don't defeat the purpose of fake mode export.

I think you are referring to an API change on FakeMode support. In current API usage, `state_dict()` is also used to imply users to save weights/buffers beforehand. The value of this PR is simply correcting the way of doing it to pick up more cases. 

I guess if we are not eager to pick up the non-persistent buffer cases (in 2.1.0), I am fine with closing this PR, and have a more well-design on including "saving weights/buffers/constant to external files".

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 4 above)
Comment (titaiwangms) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:901]: It does. `load_state_dict ` is only needed in FxSymbolicTracer. I guess dynamo somehow manages to have the info without `load_state_dict`.

Comment (BowenBao) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:901]: That's not the case I think, both are emulating how a pretrained model is loaded in pytorch. It is unrelated to dynamo. `from_pretrained` is a transformers api, while `load_state_dict` is the standard [pytorch api](https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-for-inference).

**Pull request closed without merging**